### ISSUE TITLE
maschu's summary of where we stand with the universal stationID discussion

### ISSUE THREAD

**Issue by maschu09** [openaq/project-universal-stationID#10]
1) we need something that can be automatically created based on metadata received in all data submissions
2) it appears that geographic location of a station is a reasonable proxy for the identity of a station 
3) geographic location should be encoded in an unambiguous way - plus codes appear a good way to do this
4) geographic location alone is not unambiguous, because there may be multiple facilities at one location (for example towers with different sampling heights)
5) plus code needs to be extended to encode distinguishable feature. Altitude (meaning height of sampling inlet and not geographic station altitude!) would be useful and in most cases sufficient - however this information is rarely available. We could resort to a set of defaults in this case. See #9.
6) the universal, unambiguous stationID does not aim at being recognizable by humans, but we aim at allowing established station id schemes to be coded in other metadata (as "program specific id" or "ProviderStationID".

**Comment by RocketD0g** [comment]
This is a really good summary, @maschu09 and makes a lot of sense. I think we are converging onto a Station ID format!

### My Summarized Thoughts
For the ID format, I agree with @maschu09 above, with the exception of Part 5, and instead would jump to:

> ....if sampling altitude is unknown, use letters; "A", "B", "C", etc. if there is a need to distinguish "stations"

that he outlines [here](https://github.com/openaq/project-universal-stationID/issues/9#issuecomment-463840457). Thoughts on this?

 Also tagging @sethvincent to let us know what is sounding reasonable (or unreasonable) about this conversation.

### Longer Comment



@maschu09, my one quibble would be with Point 5 above on incorporating inlet height into the ID for two reasons:

**(A)  No existing station in the system has this information currently, so all would default to the "A", "B", "C" lettering you outline [here](https://github.com/openaq/project-universal-stationID/issues/9#issuecomment-463840457) anyway.**
 
And, sticking with this being a persistent ID, we wouldn't want to change this ID, presumably, even if we do obtain station inlet height data for those stations later on. So then, even assuming a highly idealized situation in which all new stations input into the metadata system contain inlet height, this would create a dichotomy between the "pre" metadata station IDs (of which we'll have about 10300 for each station currently in the system) and the "post" metadata system. 

**(B) Inlet heights tend to be instrument-specific rather than station-specific (see below for how I'm defining a station).** 

Basically, the same little shed can contain instrumentation that measures at multiple inlet heights. The inlet height is also  easily changeable (I've unfortunately seen this happen on the fly at a station, eek) on a shed or tower, which seems not ideal for a persistent station ID.

In the current draft of the [metadata format](https://github.com/openaq/project-metadata-format/issues/7#issuecomment-463309289), the inlet height, `inletHeight` is attached to _each instrument_ at a station for this reason, while the geographic station altitude, `stationAltitude`, is attached to the station-level information. In light of this conversation, I am adjusting the [draft metadata format](https://github.com/openaq/project-metadata-format/issues/7#issuecomment-463309289) to reflect that an instrument may measure at more than one inlet height.

**I guess this ultimately comes down to defining what a station is. Here is my graphical take on what a station is:**

![stations](https://user-images.githubusercontent.com/13404290/52875482-71d5d900-3122-11e9-9866-c6e1470f4074.jpg)

Where we would capture the inlet height at the instrument level in the [metadata format](https://github.com/openaq/project-metadata-format/issues/7#issuecomment-463309289).

**Defined in words, I'd suggest a station definition as:**

- A location with defined geo coords 
- This location takes (or has taken) AQ measurements with 1 or more instruments (no matter their inlet height)

~-A  location that has been designated by a single name by the originating source (e.g. 'Anand Vihar', 'Mandir Marg', etc.)~ _EDIT: I realize this would still make us reliant on the originating source name, which is variable. Perhaps if the underlying source name changes for a given station, we could at least flag it in the metadata format._

~- This(these) instrument(s) are run by the same entity.~ _EDIT: I guess, so that we can actually automate this process, we _shouldn't_ make this a criteria, but rather again, perhaps put a flag in the metadata for when a Station ID is associated with multiple entities that are attributed to the same geographic coordinates._

If those conditions are met, it 'deserves' a unique station ID, even if there are multiple inlet heights (in which case, this should be recorded at the instrument level in the metadata format). 

~In the event that instruments are run by different entities and/or have different 'station names' as designated by the originating entity but are at the same geospatial coordinates (either due to coarseness of the coordinates or they are at different altitudes), they would get their own Station ID.~ _EDIT: Again, so that this can actually be an automated process, we probably just have to stick with coordinates, but again, this could be flagged in the metadata for when a station ID is associated with  multiple entities that are attributed to the same geographic coordinates._

 ~The one quasi-exception I see to this rule is that sometimes sources _slightly_ change the originating source name. For instance, a change like 'Anand Vihar' to 'Anand-Vihar.' These should not be considered separate stations unless other information indicates the station has moved or is under the control of a new entity. Such 'judgement calls" could be noted in the `notesByAuthor` fields in the [draft metadata format](https://github.com/openaq/project-metadata-format/issues/7#issuecomment-463309289).~ _EDIT: Disregard now, based on edits above._


Last thing I'd note is a good point that @jrybicki-jsc made [here](https://github.com/openaq/project-universal-stationID/issues/7#issuecomment-463963026): While the vast majority of stations in OpenAQ are associated with coordinates, a chunk are not. I suggest we consider them "undeserving" of station IDs until we have that basic info. Curious what @olafveerman @jflasher think on that (and of course any other thoughts on this generally).

**Comment by maschu09** [comment]
@RocketD0g Good points. Indeed the real question is what we understand under the term "station". In your little sketch, the left had drawing may actually cause issues with using coordinates as stationID: if a station has a building with some instruments and an outdoor field (which could be a couple 100 m away), the exact coordinates differ. This only matters if we insist on the notion that a station is some entity where things belong together. If we regard it only as "measurement location", then this presents no problem. Perhaps it would be cleanest then to abandon the term "station" and only use "measurement location". With this you can immediately also capture mobile measurements.
But then we don't even need a unique identifier, do we? 

Hmmmm. (scratching my head)... Let's go back to why we wanted this in the first place: the point was to have some unique identifier that remains constant even if some "station" metadata changes.  

Now, if we define a formula that generates the unique id from the given metadata, then this formula is of course vulnerable to changes in those pieces of the metadata that are used. Specifically, if the id is based on lon and lat, and the apparent location changes because of addition or removal of a decimal, our algorithm might assign a new unique id.

Thus, a location based algorithm would need to include some more logic and almost certainly some sort of manual screening for at least a subset of "stations". One could, for example, record the nearest distance to a neighbouring site, use this to define a threshold of how similar a coordinate must be to be counted as "same station", and "raise an alarm" if this distance changes. There are several possible reasons for such change: 1) a neighbour has been added or removed, 2) an error was made in the neighbouring station's coordinates, 3) an error was made in the current stations coordinates.

I am beginning to wonder if this is really something you should take on at OpenAQ or if this is actually a pure IntelliAQ/TOAR issue. Perhaps it would be more important that you try to ensure that at least coordinates are always provided  (otherwise there is little value in a measurement, is there?), and that coordinates are always given with at least 4 decimals.

In the original proposal I was thinking of a stationID that would primarily be reported by the data providers and only computed automatically where no provider id exists. Perhaps this is still the better way after all?

**Comment by RocketD0g** [comment]
I totally agree with @maschu09 in #6 on this probably being an exercise about achieving less than perfection, heh heh; I think it's the nature of this super cool, but very messy, disparately-generated data set.

I think perhaps similar to what @maschu09  said in #6, it could be useful if we think of the Station ID exercise as: 

**What's the way we can define a Station ID so that we find as many errors/concerns in the system as quickly as possible and then be able to transparently share them with the public?**

Going back to the auto-generate IDs by defining them by coordinates idea (regardless of how we actually name the IDs), we could at least then:

 a) Solve the issue of station names that have changed but are the same station (we _know_ this has happened in Delhi and currently it is definitely tricky for the user to know these are actually the same stations).

b) Flag any 'confusing' stations in the metadata format (e.g. where there appear to be more than one station at one spot, where geo coords are not precise enough, etc.). 

c) Be sure to share with the community the limitation of this approach (from how we are defining a station to limitations in what errors we might be catching or missing).

The cool thing about doing this is that we could also then enlist our community to investigate specific questions like: "Hey, in Sao Paolo, at this block, does anyone know if there actually are two physically different stations? Or did the names just change? Or does anyone suggest contacts at xxxx agency to get to the bottom of this?" and it also gives folks a place to react to the way we have presented the  information (e.g. ("I work at Station Z, which is five instruments located over the span of 3 km...this is not well represented by your conception of a station and the precision of geo coords you require.").  

Implementing this ID + metadata format can be a great way to shine a light on the messy parts of the system, engage the community to clean them up a bit, and hopefully continue to connect us to new data providers to get more seamlessly connected (or connected at all) to more data and meta data.

At any rate, we've reached our public comment period deadline now, so we'll let the Development Seed folks run with this for a bit.

**Comment by RocketD0g** [comment]
BTW: Since we're hitting COB today and the awesome Development Seed team will now be taking some time to put their heads together on all of this feedback, we just wanted to thank everyone for their contributions these past two weeks, especially, @maschu09 @jrybicki-jsc @jflasher @sethvincent @olafveerman @jflasher @urbanemissions @unorthodox-engineers @keknowland @aretey and others who relayed their comments outside of GH.
